### PR TITLE
Add healthcheck for go2rtc service

### PR DESCRIPTION
**Pull request by felipecrs**
Also:

- Upgrade s6-overlay from 3.1.3.0 to 3.1.4.0.
- Allow go2rtc to restart within the container's lifetime
- Fix startup and finish logs not being collected by s6, thus not shown in Frigate UI

### CODE REVIEW

File: docker/rootfs/etc/s6-overlay/s6-rc.d/go2rtc-healthcheck/run [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+# shellcheck shell=bash
+# Start the go2rtc healthcheck service
+
+set -o errexit -o nounset -o pipefail
+
+echo "[INFO] Starting go2rtc healthcheck service..." >&2
+
+while sleep 1m; do
+    # Check if the service is running
+    if ! curl -fsSL http://127.0.0.1:1984/api/streams >/dev/null; then
```
Comment (NickM-27) [inline docker/rootfs/etc/s6-overlay/s6-rc.d/go2rtc-healthcheck/run:11]: Is there a way to set a timeout here? I think we want specific behavior for how long curl waits until it considers it didn't get a response

File: docker/rootfs/etc/s6-overlay/s6-rc.d/go2rtc-healthcheck/run [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+        echo "[ERROR] The go2rtc is apparently not working, restarting it..." >&2
```
Comment (NickM-27) [inline docker/rootfs/etc/s6-overlay/s6-rc.d/go2rtc-healthcheck/run:12]:
```suggestion
        echo "[ERROR] The go2rtc is not responding to ping, restarting it..." >&2
```

File: docker/rootfs/etc/s6-overlay/s6-rc.d/go2rtc-healthcheck/run (hunk 1 above)
Comment (felipecrs) [reply docker/rootfs/etc/s6-overlay/s6-rc.d/go2rtc-healthcheck/run:11]: Yes, good point.

File: docker/rootfs/etc/s6-overlay/s6-rc.d/go2rtc-healthcheck/run [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+# Start the go2rtc healthcheck service
+
+set -o errexit -o nounset -o pipefail
+
+echo "[INFO] Starting go2rtc healthcheck service..." >&2
+
+while sleep 1m; do
+    # Check if the service is running
+    if ! curl -fsSL http://127.0.0.1:1984/api/streams >/dev/null; then
+        echo "[ERROR] The go2rtc is not responding to ping, restarting it..." >&2
```
Comment (NickM-27) [inline docker/rootfs/etc/s6-overlay/s6-rc.d/go2rtc-healthcheck/run:12]:
```suggestion
        echo "[ERROR] The go2rtc service is not responding to ping, restarting..." >&2
```

Comment (NickM-27) [reply docker/rootfs/etc/s6-overlay/s6-rc.d/go2rtc-healthcheck/run:12]: Sorry, wanted to tweak a bit more

**Comment by felipecrs** [comment]
For some reason this new service isn't even starting for me. I'm trying to debug what's happening.

**Comment by felipecrs** [comment]
@NickM-27, we can probably speed up the go2rtc initialization if we only create the config once. For that, I'd create the config in /dev/shm, and on next initializations I'd check if the config exists already and if it does, use it rather than recreate it.

What do you think?

**Comment by NickM-27** [comment]
> @NickM-27, we can probably speed up the go2rtc initialization if we only create the config once. For that, I'd create the config in /dev/shm, and on next initializations I'd check if the config exists already and if it does, use it rather than recreate it.
> 
> What do you think?

Yeah I think that makes sense

File: docker/rootfs/etc/s6-overlay/s6-rc.d/go2rtc-healthcheck/run [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+set -o errexit -o nounset -o pipefail
+
+echo "[INFO] Starting go2rtc healthcheck service..." >&2
+
+# Give some additional time for go2rtc to start before start pinging
+sleep 10s
+
+while sleep 1m; do
+    # Check if the service is running
+    if ! timeout 10s curl -fsSL http://127.0.0.1:1984/api/stressams >/dev/null; then
```
Comment (NickM-27) [inline docker/rootfs/etc/s6-overlay/s6-rc.d/go2rtc-healthcheck/run:14]: Is `stressams` a typo?

Comment (felipecrs) [reply docker/rootfs/etc/s6-overlay/s6-rc.d/go2rtc-healthcheck/run:14]: ops, I was forcing a failure as part of a test. lol

**Comment by felipecrs** [comment]
Ok, so this is working. If I copy and paste the healthcheck's run contents to frigate's run contents, it works great.

However, s6 is not starting the go2rtc-healthcheck service itself and I really don't understand why.

This should be the only missing thing.

**Comment by NickM-27** [comment]
> Ok, so this is working. If I copy and paste the healthcheck's run contents to frigate's run contents, it works great.
> 
> However, s6 is not starting the go2rtc-healthcheck service itself and I really don't understand why.
> 
> This should be the only missing thing.

do you need to add the `producer-for` to the go2rtc part like it is set for the logs?

**Comment by felipecrs** [comment]
I'm checking here.

Another thing, since go2rtc is now allowed to restart, we need to be more careful during initialization:

If go2rtc config for example is invalid, go2rtc will keep restarting forever. We should handle this case and make the container exit, I think.

Or maybe not?

**Comment by NickM-27** [comment]
> I'm checking here.
> 
> Another thing, since go2rtc is now allowed to restart, we need to be more careful during initialization:
> 
> If go2rtc config for example is invalid, go2rtc will keep restarting forever. We should handle this case and make the container exit, I think.
> 
> Or maybe not?

I think we should definitely fail if the create_config.py errors out, but if go2rtc doesn't like the config then that is something the create_config.py script should handle to ensure we only provide go2rtc with a valid config.

**Comment by felipecrs** [comment]
I don't think this is possible, because the create_config.py will pass whatever the user has set directly to go2rtc. Which may be a valid YAML but not a valid config.

**Comment by felipecrs** [comment]
> do you need to add the producer-for to the go2rtc part like it is set for the logs?

That fixed the issue, but I think this is probably a bug of s6-overlay. I should probably report it :(

**Comment by NickM-27** [comment]
> I don't think this is possible, because the create_config.py will pass whatever the user has set directly to go2rtc. Which may be a valid YAML but not a valid config.
> 
> i.e. we need to invoke go2rtc

actually I don't think it will be an issue. If the frigate config contains invalid yaml then frigate itself will not keep running. Personally, I think if the user provides an invalid config it is okay. It may change in the future but for now I believe go2rtc does not actually do yaml validation. For example, if you add

```
go2rtc:
  logs:
    ...
```

it should be `log` but go2rtc runs and just ignores it

**Comment by NickM-27** [comment]
> > do you need to add the producer-for to the go2rtc part like it is set for the logs?
> 
> That fixed the issue, but I think this is probably a bug of s6-overlay. I should probably report it :(

Not that I know s6 super well but I think it makes sense that it would be required. all the other ones seem to require producer-for or consumer-for

**Comment by felipecrs** [comment]
> all the other ones seem to require producer-for or consumer-for

Only because they make use of s6's logging services. I was not trying to integrate the healthcheck with the logging service, but it turns out to be a nice thing.

Now the logs from both go2rtc and the healthcheck is merged and can be seen from Frigate UI.

Also, some initialization logs were being hidden and they are now part of the logs that can be seen from Frigate UI.

File: docker/rootfs/etc/s6-overlay/s6-rc.d/go2rtc-healthcheck/run [hunk 5]
```diff
@@ -4,17 +4,18 @@
...
 
-echo "[INFO] Starting go2rtc healthcheck service..." >&2
+# Logs should be sent to stdout so that s6 can collect them
 
 # Give some additional time for go2rtc to start before start pinging
 sleep 10s
+echo "[INFO] Starting go2rtc healthcheck service..."
 
-while sleep 1m; do
+while sleep 1s; do
```
Comment (felipecrs) [inline docker/rootfs/etc/s6-overlay/s6-rc.d/go2rtc-healthcheck/run:13]: This is just a test... I'll change it.

**Comment by felipecrs** [comment]
> 

Ok, let's revisit this later if we find a need then.

**Review by NickM-27 — APPROVED** [review]
Been running this for a few minutes and all seems well, the logging improvements are nice as it should give some more info as well

**Comment by felipecrs** [comment]
Maybe we can decrease the time window to 30s?

**Comment by NickM-27** [comment]
> Maybe we can decrease the time window to 30s?

I think that would be fine, I definitely wouldn't go lower than that though. Curious what @blakeblackshear thinks

File: docker/rootfs/etc/s6-overlay/s6-rc.d/go2rtc-healthcheck/run [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+
+# Logs should be sent to stdout so that s6 can collect them
+
+# Give some additional time for go2rtc to start before start pinging
+sleep 10s
+echo "[INFO] Starting go2rtc healthcheck service..."
+
+while sleep 60s; do
+    # Check if the service is running
+    if ! timeout 10s curl -fsSL -o /dev/null http://127.0.0.1:1984/api/streams 2>&1; then
```
Comment (blakeblackshear) [inline docker/rootfs/etc/s6-overlay/s6-rc.d/go2rtc-healthcheck/run:15]: Is there a good reason to not use the builtin timeout parameters for curl?

**Comment by blakeblackshear** [comment]
30s seems fine. I don't see any reason this would add significant overhead.

File: docker/rootfs/etc/s6-overlay/s6-rc.d/go2rtc-healthcheck/run (hunk 6 above)
Comment (felipecrs) [reply docker/rootfs/etc/s6-overlay/s6-rc.d/go2rtc-healthcheck/run:15]: No good reason, but there is probably no good reason not to use bash's timeout either.

Anyway, I'll change to curl.

**Review by blakeblackshear — APPROVED** [review]

**Pull request merged**